Scripts/Weapon/DashSkill.cs

```diff
@@ -198,81 +214,70 @@ protected override void OnAttackButtonReleased()
-                // cosAngle should not be 90 degrees as that would mean that
-                // player is moving parallel to the wall hence ray would not intersect
-                targetLocation = position + hyp * GlobalTransform.basis.z;
-                if ((targetLocation - currentLocation).Length() > range)
-                    targetLocation = currentLocation + range * (-GlobalTransform.basis.z);
+                Vector3 enemyBoundary = (Vector3)rayToEnemy["position"];
+                targetLocation = enemyBoundary + player.Scale * GlobalTransform.basis.z;
             }
             else
             {
-                targetLocation = currentLocation + range * (-GlobalTransform.basis.z);
+                targetEnemy = null;
             }
         }
 
-        /*Transform currentTransform = player.GlobalTransform;
-        Transform targetTransform = new Transform(player.GlobalTransform.basis, targetLocation);
-
-        float duration = (targetLocation - currentTransform.origin).Length() / dashSpeed;
-
-        tween.InterpolateProperty(player, "GlobalTransform", currentTransform, targetTransform, duration, transType: Tween.TransitionType.Quart);
-        tween.InterpolateCallback(this, duration, "OnFinished");
-        tween.Start();
+        if (targetEnemy == null)
+        {
+            targetLocation = currentLocation + range * (-GlobalTransform.basis.z);
+        }
 
+        float duration = (targetLocation - currentLocation).Length() / dashSpeed;
         player.IsMovementLocked = true;
-        isRunning = true;*/
+        isRunning = true;
+        dashLeft = duration; //dashDuration;
         Vector3 direction = (targetLocation - currentLocation).Normalized();
         // Impulse / m = dv
-        player.ApplyCentralImpulse(direction * player.Mass * 100);
+        player.disableFriction = true;
+        player.ApplyCentralImpulse(player.Mass * -player.LinearVelocity);
+        player.ApplyCentralImpulse(direction * player.Mass * dashSpeed);
     }
 
     public void OnFinished()
     {
+        MeshInstance m = player.GetNode<MeshInstance>("Shockwave");
+        ShaderMaterial mat = m.GetSurfaceMaterial(0) as ShaderMaterial;
+        mat.SetShaderParam("t", 0);
+        // mat.GetShaderParam("TIME");
+        shaderTime = 0;
+        player.ApplyCentralImpulse(player.Mass * -player.LinearVelocity);
+
         if (targetEnemy != null)
         {
             Vector3 targetEnemyDirection = (targetEnemy.GlobalTransform.origin - GlobalTransform.origin).Normalized();
-            Vector3 accelerationOntargetEnemy = 50.0f * targetEnemyDirection + 40.0f * Vector3.Up;
-            targetEnemy.Velocity += accelerationOntargetEnemy;
+            Vector3 accelerationOntargetEnemy = 5.0f * targetEnemyDirection + 15.0f * Vector3.Up;
+            // targetEnemy.ApplyCentralImpulse(targetEnemy.Mass * -targetEnemy.LinearVelocity);
+            targetEnemy.ApplyCentralImpulse(targetEnemy.Mass * accelerationOntargetEnemy);

```

Comment:
Should this be force instead? F=ma? Otherwise not supposed to be called accelerationOnTargetEnemy?